### PR TITLE
feature: supported `redirect` plugin.

### PR DESCRIPTION
**Pull request by membphis**
fix https://github.com/iresty/apisix/issues/731

- [x] doc

### CODE REVIEW

File: lua/apisix/plugins/rewrite.lua [hunk 1]
```diff
@@ -34,7 +37,7 @@ function _M.check_schema(conf)
 end
 
 
-local function concat_new_uri(uri, ctx)
+local function parse_uri(uri)
```
Comment (moonming) [inline lua/apisix/plugins/rewrite.lua]: this function is useful for `rewirte-proxy` plugin too.

File: conf/config.yaml [hunk 2]
```diff
@@ -71,6 +71,7 @@ plugins:                          # plugin list
   - serverless-post-function
   - openid-connect
   - proxy-rewrite
+  - rewrite
```
Comment (moonming) [inline conf/config.yaml]: `redirect` is better.

File: lua/apisix/plugins/rewrite.lua [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+local function parse_uri(uri)
+    local reg = [[\$\{([0-9a-zA-Z_]+)\} | \$([0-9a-zA-Z_]+) | ([^$]+)]]
```
Comment (moonming) [inline lua/apisix/plugins/rewrite.lua]: need add some comments or examples here, not only in test cases.

File: lua/apisix/plugins/rewrite.lua [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return tab_concat(t, "")
+end
+
+
+function _M.rewrite(conf, ctx)
+    core.log.info("plugin rewrite phase, conf: ", core.json.delay_encode(conf))
+
+    local new_uri, err = concat_new_uri(conf.uri, ctx)
+    if not new_uri then
+        core.log.error("failed to genera new uri by: ", conf.uri, " error: ",
```
Comment (moonming) [inline lua/apisix/plugins/rewrite.lua]: `genera` is bad.

File: t/plugin/rewrite.t [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+repeat_each(1);
+no_long_string();
+no_shuffle();
+no_root_location();
+log_level('info');
+run_tests;
+
+__DATA__
+
+=== TEST 1: sanity
```
Comment (moonming) [inline t/plugin/rewrite.t]: not test cases for error.

Comment (membphis) [reply t/plugin/rewrite.t]: I do not know how to generate error.

**Comment by membphis** [comment]
@moonming please take a look

File: doc/README.md [hunk 6]
```diff
@@ -35,3 +35,4 @@ Plugins
 * [serverless](plugins/serverless-cn.md)：AllowS to dynamically run Lua code at *different* phase in APISIX.
 * [ip-restriction](plugins/ip-restriction.md): IP whitelist/blacklist.
 * openid-connect
+* [redirect](plugins/redirect.md): Client redirect with new uri.
```
Comment (moonming) [inline doc/README.md]: `URI redirect` is ok. `client redirect` is confused

File: doc/README_CN.md [hunk 7]
```diff
@@ -36,3 +36,4 @@ Reference document
 * [serverless](plugins/serverless-cn.md)：允许在 APISIX 中的不同阶段动态运行 Lua 代码。
 * [ip-restriction](plugins/ip-restriction-cn.md): IP 黑白名单。
 * openid-connect
+* [redirect](plugins/redirect.md): 使用新 uri 让客户端跳转。
```
Comment (moonming) [inline doc/README_CN.md]: ditto

File: t/plugin/redirect.t [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: redirect
+--- request
+GET /hello?name=json
```
Comment (moonming) [inline t/plugin/redirect.t]: what will happen if no `name=json`? need test cases to cover

File: doc/plugins/redirect.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
+[中文](redirect.md)
+
+# redirect
+
+upstream proxy info rewrite plugin.
```
Comment (moonming) [inline doc/plugins/redirect.md]: need Chinese version.
and `upstream proxy info rewrite plugin` is confused, is it copy-paste?

File: doc/plugins/redirect.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+|Name    |Required|Description|
+|------- |-----|------|
+|uri     |required| New uri which can contain Ningx variable, eg: `/test/index.html`, `$uri/index.html`. You can refer to variables in a way similar to `${xxx}` to avoid ambiguity, eg: `${uri}foo/index.html`. If you just need the original `$` character, add `\` in front of it, like this one: `/\$foo/index.html`.|
+|ret_code|option|Response code, the default value is `302`.|
+
+### Example
+
+#### Enable Plugin
+
+Here's an example, enable the `redirect` plugin on the specified route:
```
Comment (moonming) [inline doc/plugins/redirect.md]: need add more examples for `args` and `variable`.

File: t/plugin/redirect.t (hunk 8 above)
Comment (membphis) [reply t/plugin/redirect.t]: this test case has covered

https://github.com/iresty/apisix/pull/732/files/560ea19a12b3daeaa69d928e2cc3f9dc992f79ff#diff-46578132010f81fd120ed4af24c96b22R197

--- response_headers
Location: /hello/test/ajsonc

File: doc/plugins/redirect.md (hunk 9 above)
Comment (membphis) [reply doc/plugins/redirect.md]: I'll create a todo issue about the Chinese version.

File: lua/apisix/plugins/redirect.lua [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+local function concat_new_uri(uri, ctx)
+    local pased_uri_segs, err = lrucache(uri, nil, parse_uri, uri)
+    if not pased_uri_segs then
+        return nil, err
+    end
+
+    core.table.clear(tmp)
+
+    for _, uri_segs in ipairs(pased_uri_segs) do
+        local pat1 = uri_segs[1]
```
Comment (moonming) [inline lua/apisix/plugins/redirect.lua]: what's mean of `pat1` ... `pat4`? which are hard to understand.

File: lua/apisix/plugins/redirect.lua [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        local pat2 = uri_segs[2]
+        local pat3 = uri_segs[3]
+        local pat4 = uri_segs[4]
+        core.log.info(core.json.delay_encode(uri_segs))
+
+        if pat2 then
```
Comment (moonming) [inline lua/apisix/plugins/redirect.lua]: ditto

File: t/plugin/redirect.t (hunk 8 above)
Comment (moonming) [reply t/plugin/redirect.t]: I mean 
```
--- request
GET /hello?foo=json
```
not contains `name`.

File: t/plugin/redirect.t [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "plugins": {
+                        "redirect": {
+                            "uri": "$uri/test/a${arg_name}c",
```
Comment (moonming) [inline t/plugin/redirect.t]: need test case to cover`"uri": "$bad_var/test",`

File: doc/plugins/redirect.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
+[中文](redirect.md)
```
Comment (juzhiyuan) [inline doc/plugins/redirect.md]: it should be `plugins/redirect-cn.md`

File: doc/README_CN.md [hunk 15]
```diff
@@ -36,3 +36,4 @@ Reference document
 * [serverless](plugins/serverless-cn.md)：允许在 APISIX 中的不同阶段动态运行 Lua 代码。
 * [ip-restriction](plugins/ip-restriction-cn.md): IP 黑白名单。
 * openid-connect
+* [redirect](plugins/redirect.md): URI 重定向。
```
Comment (juzhiyuan) [inline doc/README_CN.md]: `plugins/redirect-cn.md`

File: lua/apisix/plugins/redirect.lua (hunk 11 above)
Comment (membphis) [reply lua/apisix/plugins/redirect.lua]: This is related to the regular content. It’s hard to take a good name here. Do you have any suggestions?

File: t/plugin/redirect.t (hunk 13 above)
Comment (membphis) [reply t/plugin/redirect.t]: did

**Comment by membphis** [comment]
Rebased this branch and added new comments, please take a look.

File: lua/apisix/plugins/redirect.lua (hunk 11 above)
Comment (membphis) [reply lua/apisix/plugins/redirect.lua]: I added some comments.

**Pull request merged**